File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-3.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        //sue 2 pointer left->dummy,right->head
        ListNode left = dummy;
        ListNode right = head;
        //shift right n times for offset
        while(n>0){
            right = right.next;
            n--;
        }
        //shift both left and right pointer until right == null
        while(right!=null){
            left = left.next;
            right = right.next;
        }
        //delete the nth node
        left.next = left.next.next;
        return dummy.next;

    }
}

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # hs = set()
        # curr = head;

        # while curr:
        #     if curr in hs:
        #         return True
        #     hs.add(curr)
        #     curr = curr.next
        # return False

        s=head
        f=head
        while f and f.next:
            s = s.next
            f = f.next.next
            if f == s:
                return True
        return False


        


        